### PR TITLE
Created EntityPersistenceNamingStrategy for replacing FileNamingStrategy

### LINKED ISSUE #246: Rename and move FileNamingStrategy
The initial idea of the the `FileNamingStrategy` has been to provide an easy-to-use mapping of Entity classes to their corresponding file names to provide a standard naming strategy. As the `FileNamingStrategy` is now used in several places that are not file related (e.g. `InfluxDbSink`) one might consider marking it as deprecated and replace it with a copy of itself but with a proper naming in a proper package e.g. `EntityPersistenceNamingStrategy` or similar. 

Whoever is starting working on this, please DO NOT DELETE the class, but mark it as deprecated and introduce the new class incl. new constructors etc. for the new class incl. an easy converter between old and new naming strategy. For an example of this pls consult me before start working.

### PR DESCRIPTION
**Pull request by ahamshubham**
Resolves #246. Changes made:

- Added deprecated tag to javadoc of `FileNamingStrategy.java`
- Created `EntityPersistenceNamingStrategy.java` as a copy of `FileNamingStrategy.java` for replacing `FileNamingStrategy.java`
- Refactored the code by changing the order of the functions for better readability

### CODE REVIEW

**Comment by ahamshubham** [comment]
!test

**Comment by ahamshubham** [comment]
- Replaced occurences of `FileNamingStrategy` with `EntityPersistenceNamingStrategy`
- Updated the javadocs to reflect the changes in the file name
- Moved `HierarchicFileNamingStrategy.java` and `HierarchicFileNamingStrategyTest.groovy` to package `edu.ie3.datamodel.io.naming`

**Comment by ahamshubham** [comment]
Sonarqube shows several warnings with respect to the regex patterns declared in `EntityPersistenceNamingStrategy.java`: 'Use the named groups of this regex or remove the names.'. Also, some Codacy Static Code Analysis issues were with respect to the size of some functions in `EntityPersistenceNamingStrategyTest.groovy`. Should I work on resolving these issues?

**Comment by ckittl** [comment]
> Sonarqube shows several warnings with respect to the regex patterns declared in `EntityPersistenceNamingStrategy.java`: 'Use the named groups of this regex or remove the names.'. Also, some Codacy Static Code Analysis issues were with respect to the size of some functions in `EntityPersistenceNamingStrategyTest.groovy`. Should I work on resolving these issues?

The code quality of tests in detail doesn't really matter to me. But if there is something in production code, we should try to fix it (cf. "camp site rule" or "scout's rule" of clean code).

**Comment by ahamshubham** [comment]
I have fixed the errors pointed out by Codacy and I tried fixing the issue pointed out by Sonarqube which says that there are unused named groups in the regex patterns in the file 'EntityPersistenceNamingStrategy.java'. To resolve the issue, I removed the function call to get the member variable with the member variable itself (since the call was being made inside the class of which the variable is a member variable). The issue still exists according to Sonarqube. The problem is that the named groups are actually being used in the same file (contradictory to what Sonarqube says) and so cannot be removed as Sonarqube points out.

**Comment by ahamshubham** [comment]
Quality gate test, however, is failing because I guess there is duplicate code as we have both `FileNamingStrategy.java` and `EntityPersistenceNamingStrategy.java` and their respective test case files in the code.

**Comment by ahamshubham** [comment]
This commit can be reverted if the files are not supposed to be deleted.

File: src/test/groovy/edu/ie3/datamodel/io/naming/EntityPersistenceNamingStrategyTest1.groovy [hunk 1]
```diff
@@ -0,0 +1,529 @@
...
+import edu.ie3.datamodel.models.timeseries.IntValue
+import edu.ie3.datamodel.models.timeseries.individual.IndividualTimeSeries
+import edu.ie3.datamodel.models.timeseries.individual.TimeBasedValue
+import spock.lang.Specification
+
+import java.nio.file.Paths
+import java.time.ZonedDateTime
+import java.util.regex.Pattern
+
+class EntityPersistenceNamingStrategyTest1 extends Specification {
```
Comment (ckittl) [inline src/test/groovy/edu/ie3/datamodel/io/naming/EntityPersistenceNamingStrategyTest1.groovy:68]: Why did you split up the test tasks? If it still is sensible, please give it a reasonable naming.

**Review by ckittl — CHANGES_REQUESTED** [review]
Looks good by around 90 % to me. Please try and improve the comments, they are mainly regarding documentation issues.

File: src/test/groovy/edu/ie3/datamodel/io/naming/EntityPersistenceNamingStrategyTest1.groovy (hunk 1 above)
Comment (ahamshubham) [reply src/test/groovy/edu/ie3/datamodel/io/naming/EntityPersistenceNamingStrategyTest1.groovy:68]: Actually, the number of functions in `EntityPersistenceNamingStrategyTest.groovy` was more than 30 due to which Codacy raised an issue. I decided to split the test cases into two files to fix the issue but I could not come up with some way in which I can decide the test case files to have a meaningful name. I will merge them both together again.

File: docs/readthedocs/io/csvfiles.rst [hunk 2]
```diff
@@ -4,25 +4,26 @@ csv files
...
-Currently we offer two different, pre-defined file naming strategies, which you might extend to fit your needs:
+A naming strategy provides a mapping between model classes and the names of the data sinks, in which the serialized
+representation of several objects of this class can be found.
+Currently we offer two different, pre-defined naming strategies, which you might extend to fit your needs:
 
-1. **FileNamingStrategy**:
-   A basic file naming strategy that is able to add prefix and suffix to the file names. A flat folder structure is
-   considered. For more details see `Default naming strategy`_.
+1. **EntityPersistenceNamingStrategy**:
+   A basic naming strategy that is able to add prefix and suffix to the names of the data sinks. A flat folder structure
```
[the remark this answers is not in the document]
Comment (ahamshubham) [reply docs/readthedocs/io/csvfiles.rst:12]: Will make the change

File: docs/readthedocs/io/csvfiles.rst [hunk 3]
```diff
@@ -4,25 +4,26 @@ csv files
...
+   An extended version of the EntityPersistenceNamingStrategy. Additionally, the `Default directory hierarchy`_ is taken
+   into account. Please note, that this directory hierarchy is only meant to be used in conjunction with input models.
 
 However, you can control the behaviour of serialization and de-serialization of models by injecting the desired naming
 strategy you like into :code:`CsvDataSource` and :code:`CsvFileSink`.
 
 Default naming strategy
 =======================
-There is a default mapping from model class to file naming in the case you would like to use csv files for
+There is a default mapping from model class to naming of data sinks in the case you would like to use csv files for
```
[the remark this answers is not in the document]
Comment (ahamshubham) [reply docs/readthedocs/io/csvfiles.rst:23]: Will make the change

File: docs/readthedocs/io/csvfiles.rst [hunk 4]
```diff
@@ -4,25 +4,26 @@ csv files
...
 (de-)serialization of models.
-You may extend / alter the naming with pre- or suffix by calling :code:`new FileNamingStrategy("prefix", "suffix")`.
+You may extend / alter the naming with pre- or suffix by calling :code:
+`new EntityPersistenceNamingStrategy("prefix","suffix")`.
```
[the remark this answers is not in the document]
Comment (ahamshubham) [reply docs/readthedocs/io/csvfiles.rst:26]: Will make the change.

File: src/main/java/edu/ie3/datamodel/io/naming/EntityPersistenceNamingStrategy.java [hunk 5]
```diff
@@ -0,0 +1,506 @@
...
+import java.util.regex.Pattern;
+import org.apache.commons.io.FilenameUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Provides an easy to use standard way to name files, tables or any other data sink that is used in
+ * order to persist entities. Normal use cases are e.g., I/O operations with .csv files or
+ * databases. If a folder structure is required for file based I/O operations, one might consider
+ * using {@link HierarchicFileNamingStrategy}
```
[the remark this answers is not in the document]
Comment (ahamshubham) [reply src/main/java/edu/ie3/datamodel/io/naming/EntityPersistenceNamingStrategy.java:37]: Will make the change.

File: src/main/java/edu/ie3/datamodel/io/naming/EntityPersistenceNamingStrategy.java [hunk 6]
```diff
@@ -0,0 +1,506 @@
...
+  }
+
+  /**
+   * Get the full path to the file with regard to some (not explicitly specified) base directory.
+   * The path does NOT start or end with any of the known file separators or file extension.
+   *
+   * @param cls Targeted class of the given file
+   * @return An optional sub path to the actual file
+   */
+  public Optional<String> getFilePath(Class<? extends UniqueEntity> cls) {
```
[the remark this answers is not in the document]
Comment (ahamshubham) [reply src/main/java/edu/ie3/datamodel/io/naming/EntityPersistenceNamingStrategy.java:159]: Okay, so I do not make any changes for this then.

File: src/main/java/edu/ie3/datamodel/io/naming/EntityPersistenceNamingStrategy.java [hunk 7]
```diff
@@ -0,0 +1,506 @@
...
+    return Optional.of(buildResultEntityString(resultEntityClass));
+  }
+
+  /**
+   * Get the the file name for all {@link AssetTypeInput}s
+   *
+   * @param typeClass the asset type class a filename string should be generated from
+   * @return the filename string
+   */
+  public Optional<String> getTypeFileName(Class<? extends AssetTypeInput> typeClass) {
```
[the remark this answers is not in the document]
Comment (ahamshubham) [reply src/main/java/edu/ie3/datamodel/io/naming/EntityPersistenceNamingStrategy.java:230]: Will make the change.

File: src/main/java/edu/ie3/datamodel/io/naming/EntityPersistenceNamingStrategy.java [hunk 8]
```diff
@@ -0,0 +1,506 @@
...
+   * @param cls Targeted class of the given file
+   * @return The file name
+   */
+  public Optional<String> getFileName(Class<? extends UniqueEntity> cls) {
+    Optional<String> inputEntityFileName = getInputEntityFileName(cls);
+    if (inputEntityFileName.isPresent()) return inputEntityFileName;
+    if (ResultEntity.class.isAssignableFrom(cls))
+      return getResultEntityFileName(cls.asSubclass(ResultEntity.class));
+    logger.error("There is no naming strategy defined for {}", cls.getSimpleName());
+    return Optional.empty();
```
[the remark this answers is not in the document]
Comment (ahamshubham) [reply src/main/java/edu/ie3/datamodel/io/naming/EntityPersistenceNamingStrategy.java:185]: Thank you. Will make the change.

File: src/main/java/edu/ie3/datamodel/io/naming/EntityPersistenceNamingStrategy.java [hunk 9]
```diff
@@ -0,0 +1,506 @@
...
+  }
+
+  /**
+   * Get the the file name for all {@link ResultEntity}s
+   *
+   * @param resultEntityClass the result entity class a filename string should be generated from
+   * @return the filename string
+   */
+  public Optional<String> getResultEntityFileName(Class<? extends ResultEntity> resultEntityClass) {
```
[the remark this answers is not in the document]
Comment (ahamshubham) [reply src/main/java/edu/ie3/datamodel/io/naming/EntityPersistenceNamingStrategy.java:220]: Will make the change.

File: src/main/java/edu/ie3/datamodel/io/naming/EntityPersistenceNamingStrategy.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (ckittl) [reply src/main/java/edu/ie3/datamodel/io/naming/EntityPersistenceNamingStrategy.java]: Thanks for pointing at this. Actually, the instances of `CharacteristicInput` have no need to be persisted anymore. Therefore, I deleted the parts, that caused the issues.

**Review by ckittl — DISMISSED** [review]

**Review by ckittl — APPROVED** [review]

**Pull request merged**